circuits/cpp/barretenberg/cpp/src/barretenberg/honk/instance/instance.hpp

```diff
@@ -0,0 +1,109 @@
+#pragma once
+#include "barretenberg/honk/flavor/goblin_ultra.hpp"
+#include "barretenberg/honk/flavor/ultra.hpp"
+#include "barretenberg/honk/flavor/ultra_grumpkin.hpp"
+#include "barretenberg/honk/proof_system/folding_result.hpp"
+#include "barretenberg/proof_system/composer/composer_lib.hpp"
+#include "barretenberg/proof_system/flavor/flavor.hpp"
+#include "barretenberg/proof_system/relations/relation_parameters.hpp"
+#include "barretenberg/srs/factories/file_crs_factory.hpp"
+#include <cstddef>
+#include <memory>
+#include <utility>
+#include <vector>
+namespace proof_system::honk {
+/**
+ * @brief  An Instance is normally constructed from a finalised circuit and it's role is to compute all the polynomials
+ * involved in creating a proof and, if requested, the verification key.
+ * In case of folded Instance, this will be created from the FoldingResult, the aggregated work from the folding prover
+ * and verifier. More specifically, a folded instance will be constructed from the complete set of folded polynomials
+ * and folded public inputs and its FoldingParams are expected to be non-zero
+ *
+ */
+
+template <UltraFlavor Flavor> class Instance_ {
+  public:
+    using Circuit = typename Flavor::CircuitBuilder;
+    using ProvingKey = typename Flavor::ProvingKey;
+    using VerificationKey = typename Flavor::VerificationKey;
+    using CommitmentKey = typename Flavor::CommitmentKey;
+    using FF = typename Flavor::FF;
+    using FoldingParameters = typename Flavor::FoldingParameters;
+    using ProverPolynomials = typename Flavor::ProverPolynomials;
+
+    // offset due to placing zero wires at the start of execution trace
+    static constexpr size_t num_zero_rows = Flavor::has_zero_row ? 1 : 0;
+
+    static constexpr std::string_view NAME_STRING = "UltraHonk";
+    static constexpr size_t NUM_WIRES = Circuit::NUM_WIRES;
+    std::shared_ptr<ProvingKey> proving_key;
+    std::shared_ptr<VerificationKey> verification_key;
+    std::shared_ptr<CommitmentKey> commitment_key;
+
+    ProverPolynomials prover_polynomials;
+
+    // The number of public inputs has to be the same for all instances because they are
+    // folded element by element.
+    std::vector<FF> public_inputs;
```

Comment:
Also we can't have the case where instance A has 50 inputs and instance B has 3 inputs and from index 4 to 50 instance B has zeros because, on one hand this is wasteful and on the other hand the number of actual public inputs matter for the compute_public_input_delta function